LSPS4/README.md

```diff
@@ -0,0 +1,1923 @@
+> reach the payee, until it gets an error from the
+> payee, indicating that it was able to successfully
+> reach the payee.
+> If the payee was a client of an LSP, and they already
+> have a channel, then there is no problem, the client
+> simply fails the payment.
+> However, if the client arranged for a JIT channel,
+> but the PPP mechanism did not exist, then the LSP
+> has to first open a channel and then send the HTLC
+> via the channel, and then the client would fail
+> the HTLC due to not knowing the preimage to the
+> randomly-generated hash, resulting in a channel
+> open that is not paid for.
+> *With* the PPP mechanism, the LSP allows the client
+> to access the HTLC details, and the client can
+> fail the HTLC using an error originating from the
+> client, which signals to the payer that it was
+> able to reach the payee, but the LSP does not
+> open a channel yet, preventing a channel open
+> that is not paid for.
+
+### 3. Pending Payment Part Notification
+
+A "pending payment part" or "PPP" is an abstraction.
+As of this specification version, all PPPs are HTLCs, but in the
+future, a PPP may be a pending asynchronous receive, or a PTLC.
+
+Every PPP has a unique string identifier, the "PPP identifier".
+The format of this identifier is defined by the LSP, but MUST be a
+high-entropy identifier, with at least 80 bits of entropy.
+The PPP identifier:
+
+* MUST be generated with at least 80 bits of entropy.
+* MAY have any format, provided:
+  * MUST be encodable in an ASCII encoded JSON string without any
+    string escapes needed.
+  * MUST not be longer than 24 bytes (not including the `"`
+    delimiters) when encoded as an ASCII-encoded JSON string.
+
+Whenever the LSP receives an incoming payment that goes to a reserved
+SCID, it:
+
+* Generates a fresh PPP identifier for it.
+* Starts a hold timer for that incoming payment, duration depending on
+  its PPP type.
+  * Does not respond to the inbound until after the hold timer expires,
+    or the client has handled the payment fully.
+* Adds the PPP to a called the "PPP Map", which maps the PPP identifier
+  to the information the LSP needs to forward, handle, or fail the
+  underlying payment, as well as the information that the client needs to
+  process for that PPP.
+  * If the hold timer expires before the client has decided to forward,
+    fail, or use the PPP for opening a new channel, should also remove
+    it from the PPP Map.
+* Sends the `lsps4.you_have_new_ppp` notification to the client, containing
+  the PPP identifier and the details the client needs to process the
+  PPP.
+
+The LSP SHOULD persist the PPP Map on-disk, so that if the LSP crashes and
+then restarts quickly, the client can reconnect and re-download the PPP Map
+from the LSP, improving payment experience.
+
+However, the LSP MAY choose not to persist the PPP Map.
+In this case, if the LSP crashes and restarts, it MUST fail all HTLCs that
+were held in PPPs with `temporary_channel_failure`.
```

Comment:
Assuming the LSP is able to persist the PPP Map somehow.

In LDK, on restart any intercepted HTLCs are failed automatically with no way for the LSP code to re-send to the client.  So we need to specify this (LDK-based LSPs cannot actually store the PPP Map).

(This is the main reason why we talk about the PPP Map, some software can recover on restart, some (LDK) cannot)

Reply:
In LND the htlcs are replayed when reconnecting to the grpc stream.
In lspd (breez lsp implementation), we replay htlcs from CLN as well when lspd reconnects.

This allows for a 'stateless' handling of these htlcs. When the node restarts, lspd still has the same state, and ignores replayed htlcs. When lspd restarts, the htlcs are replayed and lspd can continue without remembering the previous state. 

So maybe just change MUST to MAY here, in case the LSP can recover after crash.